src/dc_imex.rs

```diff
@@ -187,83 +188,78 @@ pub unsafe fn dc_initiate_key_transfer(context: &Context) -> *mut libc::c_char {
-                        CString::new(context.stock_str(StockMessage::AcSetupMsgSubject).as_ref())
-                            .unwrap();
-                    let setup_message_body = context.stock_str(StockMessage::AcSetupMsgBody);
-                    let msg_body_head: &str = setup_message_body.split('\r').next().unwrap();
-                    let msg_body_html = CString::new(msg_body_head.replace("\n", "<br>")).unwrap();
-                    ret_setupfilecontent =
-                        dc_mprintf(b"<!DOCTYPE html>\r\n<html>\r\n<head>\r\n<title>%s</title>\r\n</head>\r\n<body>\r\n<h1>%s</h1>\r\n<p>%s</p>\r\n<pre>\r\n%s\r\n</pre>\r\n</body>\r\n</html>\r\n\x00"
-                                   as *const u8 as *const libc::c_char,
-                                   setup_message_title.as_ptr(),
-                                   setup_message_title.as_ptr(),
-                                   msg_body_html.as_ptr(),
-                                   encr_string);
-                    free(encr_string as *mut libc::c_void);
-                }
-            }
-        }
+pub fn dc_render_setup_file(context: &Context, passphrase: &str) -> Result<String> {
+    ensure!(
+        passphrase.len() >= 2,
+        "Passphrase must be at least 2 chars long."
+    );
+    unsafe {
+        ensure!(
+            !(dc_ensure_secret_key_exists(context) == 0),
+            "No secret key available."
+        );
     }
-    sqlite3_finalize(stmt);
-
-    ret_setupfilecontent
+    let self_addr = context
+        .get_config(Config::ConfiguredAddr)
+        .ok_or(format_err!("Failed to get self address."))?;
+    let private_key = Key::from_self_private(context, self_addr, &context.sql)
+        .ok_or(format_err!("Failed to get private key."))?;
+    let ac_headers = match context
+        .sql
+        .get_config_int(context, Config::E2eeEnabled)
+        .unwrap_or(1)
+    {
+        0 => None,
+        _ => Some(("Autocrypt-Prefer-Encrypt", "mutual")),
+    };
+    let private_key_asc = private_key.to_asc(ac_headers);
+    let encr = {
+        let private_key_asc_c = CString::yolo(private_key_asc);
+        let passphrase_c = CString::yolo(passphrase);
+        dc_pgp_symm_encrypt(
+            passphrase_c.as_ptr(),
+            private_key_asc_c.as_ptr() as *const libc::c_void,
+            private_key_asc_c.as_bytes().len(),
+        )
+        .ok_or(format_err!("Failed to encrypt private key."))?
+    };
+    let replacement = format!(
+        concat!(
+            "-----BEGIN PGP MESSAGE-----\r\n",
+            "Passphrase-Format: numeric9x4\r\n",
+            "Passphrase-Begin: {}"
+        ),
+        &passphrase[..2]
+    );
+    let pgp_msg = encr.replace("-----BEGIN PGP MESSAGE-----", &replacement);
+
+    let msg_subj = context.stock_str(StockMessage::AcSetupMsgSubject);
+    let msg_body = context.stock_str(StockMessage::AcSetupMsgBody);
+    // let msg_body_head: &str = msg_body.split('\r').next().unwrap();
```

Comment:
@r10s I converted this previously to the line shown here, this splits the string at `\r` and takes the first part only.  IIRC this is what the original C code did.  Here I'm removing it because I have no idea what it's purpose is, the fallback stock string of AcSetupMsgBody does not contain any `\r` character, I'm not sure about translations but if one sneaks in it seems more like an obscure feature which would disappear half of a translation.

Maybe the intention was to strip `\r` occurrences so that only `\n` was kept which is than replaced with `<br>` below?  In that case we'd have to keep stripping this here as that makes much more sense.  What's your opinion?

Reply:
@flub i do not think, this is the same the c-core did.

the c-code used the _whole_ string, not only the part before the first `\r`. however, the c-code was generous to different lineends, assuming the stock-string was provided by the ui that does not really take care about that. for this purpose, the [c-code removes all `\r` and replaces all `\n` by a `<br>`](https://github.com/deltachat/deltachat-core/blob/master/src/dc_imex.c#L144) to create html-text.

so, i think the line `let msg_body_html = msg_body.replace("\n", "<br>");` should be replaces by 

```
let msg_body_html = msg_body.replace("\r", "");
let msg_body_html = msg_body.replace("\n", "<br>");
```

(or similar :)

Reply:
Thanks, confirms what I expected.  It probably was my failure to read the converted C in an earlier PR.  I fixed it and added a unittest.